enhancements/olm/olm-managed-operator-metrics.md

```diff
@@ -0,0 +1,145 @@
+
+- A fresh install for this operator is being created where this field is set to `true`
+- An existing operator had this field set to `false` and is being upgraded to a new version that sets this field to `true`
+
+> Note: The `InstallPlan` will include a helpful message for Cluster Admins if either of the situations above are encountered.
+
+### OLM Telemeter Support
+
+#### Telemeter Requirements
+
+Once OLM is capable of managing `ServiceMonitor` and `PrometheusRule` objects along with the lifecycle of an operator, OLM must allow Red Hat owned operators to take advantage of Telemeter. Before moving forward, let's review the requirements that must be met for Telemeter to scrape an application's metrics:
+
+1. The application must be deployed in a namespace that is both prefixed with `openshift-` and labeled with `openshift.io/cluster-monitoring=true`
+2. A [ServiceMonitor object](https://github.com/coreos/prometheus-operator/blob/master/Documentation/user-guides/getting-started.md#related-resources) must be created in the namespace mentioned above and point to the application's metrics endpoint
+3. A role/rolebinding must be created that provides the Prometheus Operator ServiceAccount with the appropriate RBAC priviledges to discover the `ServiceMonitors` and `PrometheusRules` objects in the newly created namespace.
+
+Both requirements 1 and 2 are fully met by the changes proposed above leaving our team with the duty to fulfill requirement 3.
+
+#### Fulfilling Requirement 3: Updating the Prometheus Operator ServiceAccount
+
+In order for the Prometheus Operator to recognize take advantage of this new feature, its ServiceAccount must have the proper RBAC permissions to view events on the `ServiceMonitors` and `PrometheusRules` objects in the operator's namespace.
+
+On upstream clusters, there is no way for OLM to identify which Prometheus Operator ServiceAccount should be granted the RBAC priviledges. As such, the cluster admin will need to update the correct Prometheus Operator's OperatorGroup to include the correct namespace.
+
+On OpenShift cluster, OLM can simply update the default Prometheus instance with the correct RBAC priviledges when the `openshift.io/cluster-monitoring=true` label is present on the namespace.
+
+#### Risks and Mitigations
+
+It is possible that the Prometheus team will not allow OLM managed operators to use the default prometheus operator to record metrics. If they are concerned about OLM managed operators overloading the prometheus instance, one solution that addresses this concern includes creating a new operator that is responsible for adding the `openshift.io/cluster-monitoring=true` label to namespaces and for generating the Telemter RBAC. This operator would:
```

Comment:
> It is possible that the Prometheus team will not allow OLM managed operators to use the default prometheus operator to record metrics. 

Openshift workload operators installed via OLM are fine to use the "default prometheus operator" e.g. the prom-operator that is used for openshift workloads. For other operators installed via OLM they are treated as user workload monitoring, and with that coming as preview from 4.3, it can be easily used in the same way, just not in openshift namespaces.

Hopefully that makes sense?

Reply:
@LiliC I want to make sure I understood what you stated above.  Any operator that installs in openshift-* namespace is eligible for inclusion in cluster-monitoring, any operator that installs outside of openshift-* namespace is treated as workload monitoring.  This means for things like IBM Cloud Paks which install in ibm-*, it would always be treated as workload monitoring, correct?

Reply:
> Any operator that installs in openshift-* namespace is eligible for inclusion in cluster-monitoring

Should have added, and that namespace has the `openshift.io/cluster-monitoring=true` label is what we consider openshift workload monitoring.

> any operator that installs outside of openshift-* namespace is treated as workload monitoring.

@derekwaynecarr any operator outside of openshift-* namespaces is treated as [user workload monitoring](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/user-workload-monitoring.md), and if this is enabled, this is available from 4.3 onwards. Before that, or if user workload monitoring is not enabled, there is no monitoring of those namespaces.

> things like IBM Cloud Paks which install in ibm-*, it would always be treated as workload monitoring, correct?

We have nothing special in place for ibm-* namespaces, that I am aware of (cc @s-urbaniak @brancz ). Right now we would treat them as **user** workload monitoring yes.

Reply:
> We have nothing special in place for ibm-* namespaces, that I am aware of (cc @s-urbaniak @brancz ). Right now we would treat them as user workload monitoring yes.

Correct, there is no special handling for ibm-* prefixed namespaces right now. All namespaces which do not start with openshift-* AND do not have the `openshift.io/cluster-monitoring=true` label set are treated as user workload monitoring as @LiliC explained.

As this requirement was also requested in another thread I created an epic https://issues.redhat.com/browse/MON-901 to track it.

Reply:
Based on reading https://github.com/openshift/cluster-monitoring-operator/blob/e1b5cf1917133907e3a74e53db3b637e48d384a5/pkg/client/client.go#L248-L262, I think I can have namespace "ibm-fobber" with `openshift.io/cluster-monitoring=true` and get added to cluster monitoring.

Reply:
@deads2k unfortunately it turned out we do have a spot with hard-coded namespace prefixes: https://github.com/openshift/cluster-monitoring-operator/blob/d531ecb19943f445a3b8ed46d593dec4fd05674a/jsonnet/main.jsonnet#L132, hence the rationale for creating the https://issues.redhat.com/browse/MON-901 epic to allow CMO to really _just_ listen to namespace monitoring labels, also for alerting.